BParkingNano/plugins/TrackMerger.cc

```diff
@@ -53,127 +57,120 @@ class TrackMerger : public edm::global::EDProducer<> {
-  int nLostTracks = lostTracks->size();      
-  int totalTracks = nTracks + nLostTracks;
+ // do we need all those ints?
+//  int nLostTracks = lostTracks->size();      
+//  int totalTracks = nTracks + nLostTracks;
+ //same as before comment out untill we are sure 
+//  std::unique_ptr<pat::CompositeCandidateCollection> outTag(new pat::CompositeCandidateCollection());
+
+//ok this was CompositeCandidateCollection 
+  std::unique_ptr<pat::CompositeCandidateCollection> tracks_out(new pat::CompositeCandidateCollection);
+
+
+//correct logic but a bit convoluted -> changing to smthn simpler
+ std::vector<pat::PackedCandidate> totalTracks(*tracks);
+
+ totalTracks.insert(totalTracks.end(),lostTracks->begin(),lostTracks->end());
+
+ 
+ // for loop is better to be range based - especially for large ensembles  
+ for( const pat::PackedCandidate & trk: totalTracks){
+
+   //arranging cuts for speed
+   if(!trk.hasTrackDetails()) continue;
+   if(abs(trk.pdgId()) != 211) continue; //do we want also to keep muons?
+   if(trk.pt() < trkPtCut_ ) continue;
+   if(fabs(trk.eta()) > trkEtaCut_) continue;
+
+   if( (trk.pseudoTrack().normalizedChi2() < trkNormChiMin_ &&
+        trkNormChiMin_>=0 ) ||
+       (trk.pseudoTrack().normalizedChi2() > trkNormChiMax_ &&
+        trkNormChiMax_>0)  )    continue; 
+
+   bool skipTrack=true;
+   GlobalPoint trgvtx;
+   for (const pat::Muon & mu: *trgMuons){
+    //remove tracks inside trg muons jet
+    if(reco::deltaR(trk, mu) < drTrg_Cleaning_ && drTrg_Cleaning_ >0) 
+      continue;
+    //if dz is negative it is deactivated
+    if((fabs(trk.vz() - mu.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ > 0))
+       continue;
+    skipTrack=false;
+    trgvtx=GlobalPoint(mu.vx(),mu.vy(),mu.vz());
+    break; // at least for one trg muon to pass this cuts
+   }
+   // if track is closer to at least a triggering muon keep it
+   if (skipTrack) continue;
+
+   // high purity requirment applied only in packedCands
+   unsigned int itrk=&trk-&totalTracks[0];
+   if( itrk < nTracks && !trk.trackHighPurity()) continue;
+   
+   //distance closest approach in x,y wrt triggeringMuon
+   std::pair<double,double> DCA = computeDCA(trk, bFieldHandle, trgvtx);
+   float DCABS = DCA.first;
+   float DCABSErr = DCA.second;
+   float DCASig = DCABS/DCABSErr;
+   if (DCASig >  dcaSig_  && dcaSig_ >0) continue;
+   pat::CompositeCandidate pcand;
+   
+   pcand.setP4(trk.p4());
+   pcand.setCharge(trk.charge());
+   pcand.setVertex(trk.vertex());
+   pcand.addUserInt("isPacked", (itrk < nTracks) ? 1 : 0);
+   pcand.addUserInt("isLostTrk", (itrk < nTracks) ? 0 : 1);      
+   pcand.addUserFloat("dxy", trk.dxy());
+   pcand.addUserFloat("dxyS", trk.dxy()/trk.dxyError());
+   pcand.addUserFloat("dz", trk.dz()); 
+   pcand.addUserFloat("dzS", trk.dz()/trk.dzError());
+   pcand.addUserFloat("DCASig", DCASig);
+   //adding the candidate in the composite stuff for fit (need to test)
+//   pcand.addUserCand("cand",trk.sourceCandidatePtr(0));
```

Comment:
We should also store the pointer to the PackedCandidate, if we want to get the TransientTrack later on

Reply:
Sure this i think is solved already ;)